delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala

```diff
@@ -0,0 +1,1169 @@
+
+        // When the target table is empty, and the optimizer optimized away the join entirely
+        // numSourceRows will be incorrectly 0.
+        // We need to scan the source table once to get the correct
+        // metric here.
+        if (cmd.metrics("numSourceRows").value == 0 &&
+          (dataSkippedFiles.isEmpty || targetDF.take(1).isEmpty)) {
+          val numSourceRows = sourceDF.count()
+          cmd.metrics("numSourceRows").set(numSourceRows)
+        }
+
+        // Update metrics
+        cmd.metrics("numTargetFilesBeforeSkipping") += deltaTxn.snapshot.numOfFiles
+        cmd.metrics("numTargetBytesBeforeSkipping") += deltaTxn.snapshot.sizeInBytes
+        val (afterSkippingBytes, afterSkippingPartitions) =
+          totalBytesAndDistinctPartitionValues(dataSkippedFiles)
+        cmd.metrics("numTargetFilesAfterSkipping") += dataSkippedFiles.size
+        cmd.metrics("numTargetBytesAfterSkipping") += afterSkippingBytes
+        cmd.metrics("numTargetPartitionsAfterSkipping") += afterSkippingPartitions
+        val (removedBytes, removedPartitions) =
+          totalBytesAndDistinctPartitionValues(touchedAddFiles)
+        cmd.metrics("numTargetFilesRemoved") += touchedAddFiles.size
+        cmd.metrics("numTargetBytesRemoved") += removedBytes
+        cmd.metrics("numTargetPartitionsRemovedFrom") += removedPartitions
+        touchedAddFiles
+      }
+
+    private def writeModified(processedDF: DataFrame): Seq[FileAction] = {
+      val outputDF = processedDF
+        .filter(!col(ROW_DROPPED_COL)) // Filter out dropped rows
+        .drop(ROW_DROPPED_COL, ROW_MATCHED_COL, METADATA_ROW_IDX_COLUMN, FILE_PATH_COL)
+
+      val repartitionedDF = cmd.repartitionIfNeeded(spark,
+        outputDF,
+        deltaTxn.metadata.partitionColumns)
+
+      deltaTxn.writeFiles(repartitionedDF)
+    }
+
+    private def writeUnmodified(processedDF: DataFrame): Seq[FileAction] = {
+
+      // We filter out the modified target rows ids and file paths
+      val modifiedTargetRows = processedDF
+        .filter(col(METADATA_ROW_IDX_COLUMN).isNotNull) // Filter out unmatched source rows
+        .select(
+          (col(METADATA_ROW_IDX_COLUMN) * -1).as(METADATA_ROW_IDX_COLUMN),
+          col(FILE_PATH_COL))
+
+      val baseTargetDF = targetDF.drop(TARGET_ROW_PRESENT_COL)
+
+      val unModifiedTargetRows = {
+        val partitionedTargetDF = addRepartitionByFilePath(baseTargetDF)
+        val df = makeUnmodifiedTargetDF(partitionedTargetDF, modifiedTargetRows)
+
+        // The correctness of RapidsRepartitionByFilePath relies on the fact we can enable
+        // FORCE_ONE_FILE_PER_PARITION in file scan, which currently only works for
+        // GpuFileSourceScanExec, so we have to rollback to original input if for some reason we
+        // can use GpuFileSourceScanExec.
+        val gpuFileScanOverride = checkGpuFileSourceScanExecOverride(df)
+
+        if (!gpuFileScanOverride) {
```

Comment:
The test falls back the new custom repartition exec (which I'm confused how this works since the CPU version of this exec does not allow execution).  Seems like the test should test what is actually expected to happen in practice (i.e.: the scan falling back) rather than the custom exec that no user knows about falling back.